File: weknot_android/app/src/main/java/com/example/weknot_android/base/BaseViewModel.java
package com.example.weknot_android.base;

import android.app.Application;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.weknot_android.room.repository.TokenRepository;
import com.example.weknot_android.room.repository.RoomRepository;
import com.example.weknot_android.room.sharedpreference.Token;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;

public abstract class BaseViewModel<DT, RT, ET> extends AndroidViewModel {

    private CompositeDisposable disposable;
    private TokenRepository tokenManager;
    private RoomRepository repository;

    protected final MutableLiveData<String> successMessage = new MutableLiveData<>();
    protected final MutableLiveData<String> errorMessage = new MutableLiveData<>();
    final MutableLiveData<DT> data = new MutableLiveData<>();
    final MutableLiveData<ET> entity = new MutableLiveData<>();
    public final MutableLiveData<RT> request = new MutableLiveData<>();

    public LiveData<String> getSuccessMessage() {
        return successMessage;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    public LiveData<DT> getData() {
        return data;
    }

    public LiveData<ET> getEntity() {
        return entity;
    }

    public Token getToken() {
        return tokenManager.getToken();
    }

    public RoomRepository getRepository() {
        return repository;
    }

    protected BaseViewModel(Application application) {
        super(application);
        disposable = new CompositeDisposable();
        tokenManager = new TokenRepository(application);
        repository = new RoomRepository(application);
    }

    public void addDisposable(Single single, DisposableSingleObserver observer) {
        disposable.add((Disposable) single.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread()).subscribeWith(observer));
    }

    public DisposableSingleObserver<String> getBaseObserver() {
        return new DisposableSingleObserver<String>() {
            @Override
            public void onSuccess(String s) {
                successMessage.setValue(s);
            }

            @Override
            public void onError(Throwable e) {
                errorMessage.setValue(e.getMessage());
            }
        };
    }

    public DisposableSingleObserver<DT> getDataObserver() {
        return new DisposableSingleObserver<DT>() {
            @Override
            public void onSuccess(DT t) {
                data.setValue(t);
            }

            @Override
            public void onError(Throwable e) {
                errorMessage.setValue(e.getMessage());
            }
        };
    }

    public DisposableSingleObserver<ET> getEntityObserver() {
        return new DisposableSingleObserver<ET>() {
            @Override
            public void onSuccess(ET t) {
                entity.setValue(t);
            }

            @Override
            public void onError(Throwable e) {
                errorMessage.setValue(e.getMessage());
            }
        };
    }
}

File: weknot_android/app/src/main/java/com/example/weknot_android/base/BaseViewModel.kt
package com.example.weknot_android.base

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import com.example.weknot_android.model.repository.RoomRepository
import com.example.weknot_android.model.repository.TokenRepository
import com.example.weknot_android.model.repository.UserIdRepository
import io.reactivex.Single
import io.reactivex.SingleObserver
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.observers.DisposableSingleObserver
import io.reactivex.schedulers.Schedulers
import java.lang.ref.WeakReference

abstract class BaseViewModel<D, N> protected constructor(application: Application) : AndroidViewModel(application) {
    private val disposable: CompositeDisposable = CompositeDisposable()
    private val tokenManager: TokenRepository = TokenRepository(application)
    private val userIdManager: UserIdRepository = UserIdRepository(application)
    protected val repository: RoomRepository = RoomRepository(application)

    private lateinit var navigator: WeakReference<N>

    var token: String
        get() = tokenManager.token.token
        set(value) = tokenManager.setToken(value)
    var userId: String
        get() = userIdManager.userId.id
        set(value) = userIdManager.setUserId(value)
    fun getNavigator(): N {
        return navigator.get()!!
    }
    fun setNavigator(navigator: N) {
        this.navigator = WeakReference(navigator)
    }

    fun addDisposable(single: Single<*>, observer: DisposableSingleObserver<*>) {
        disposable.add(single.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread()).subscribeWith(observer as SingleObserver<Any>) as Disposable)
    }

    val baseObserver: DisposableSingleObserver<String>
        get() = object : DisposableSingleObserver<String>() {
            override fun onSuccess(s: String) {
                onRetrieveBaseSuccess(s)
            }

            override fun onError(e: Throwable) {
                onRetrieveError(e)
            }
        }

    val dataObserver: DisposableSingleObserver<D>
        get() = object : DisposableSingleObserver<D>() {
            override fun onSuccess(t: D) {
                onRetrieveDataSuccess(t)
            }

            override fun onError(e: Throwable) {
                onRetrieveError(e)
            }
        }

    protected abstract fun onRetrieveDataSuccess(data: D)
    protected abstract fun onRetrieveBaseSuccess(message: String)
    protected abstract fun onRetrieveError(throwable: Throwable)
}